src/shutdown-handler.ts

```diff
@@ -13,13 +18,23 @@ async function startShutdown() {
     return;
   }
   isShuttingDown = true;
+  const timeoutError = Symbol();
   let errorEncountered = false;
-  for (const handler of shutdownHandlers) {
+  for (const config of shutdownConfigs) {
     try {
-      await Promise.resolve(handler());
+      logger.info(`Closing ${config.name}...`);
+      await resolveOrTimeout(Promise.resolve(config.handler()), 5000, timeoutError);
+      logger.info(`${config.name} closed`);
     } catch (error) {
       errorEncountered = true;
-      logError('Error running shutdown handler', error);
+      if (error === timeoutError) {
+        logError(`Closing of ${config.name} timed out`);
+        if (config.timeoutHandler) {
```

Comment:
Maybe the `else` for this condition (`if (config.timeoutHandler === null) { ... }`) should be something like:

```ts
logError(`Critical process has taken longer than ${timeoutConstant} to shutdown, possibly hanging indefinitely`);
```